enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+
+Expected symptoms of failure include:
+- An inability to switch a resource authoritative API
+- Stale data in the non-authoritative API resources
+
+None of the symptoms above impact the day to day operation of a cluster.
+They only impact administrator actions to migrate between the Machine API and Cluster API.
+
+Importantly, Machine scaling should be unaffected by the failure of the synchronisation controller.
+
+Given this, the severity of these failures is low.
+
+Stale data may be observed by fetching the non-authoritative API resource and comparing the value of the `machine.openshift.io/synchronised-generation`
+annotation with the generation of the authoritative resource.
+Errors in conversion will be reported on the Machine API resource in the `status.conditions`.
+
+The new webhook introduced by this enhancement will have `FailPolicy=Ignore`, and therefore should not impact cluster operations when failing.
+
+When the controller and webhook are restored, they should, unless there are persistent errors in need of human intervention, continue reconciling and restore the desired state.
+
+## Implementation History
+
+- [ ] Enhancement merged (YYYY-MM-DD)
+
+## Alternatives
+
+### Use an aggregated API server to handle Machine API and Cluster API
+
+As an alterative to synchronising resources post update, we could use an aggregated API server to implement both Machine API and Cluster API endpoints going forward.
+
+This aggregated API server would be responsible for handling all API calls for both API groups and would interact directly with etcd for storage of the resources.
+
+In this case, it could ensure conversion is synchronised and written to etcd atomically, in a single call.
+This would prevent any drift between resources and mean a number of the rules above would no longer be required.
+
+This alternative has been dismissed for a number of reasons:
+- Aggregated API servers are generally complex and introduce new failure modes for the entire cluster
+- We are not certain that it is possible to take over the API endpoints of an existing installed CRD
+- Interacting directly at the storage layer is complex and adds other risks that we do not want to accept
+- Handling errors writing multiple objects to etcd for a single API call becomes complex
+- We still need a way to decide which operator is authoritative
+
+### Machine API controllers write out Cluster API resources
+
+In this alternative, rather than converting between resources in a separate controller, the implementation of each Machine controller for Machine API would be rewritten.
+Instead of interacting directly with the cloud provider, the rewrite would convert the resources from Machine API to Cluster API and then write these to the API server.
+
+The controllers would become shims, leveraging Cluster API as a backend to implement their existing functionality.
+
+This alternative has been dismissed based on the following reasons:
```

Comment:
We are still likely to use this model for OpenStack anyway. I think it is reasonable to make this decision per-provider, and it should be encouraged where possible. It may not be possible or practical for all providers.

The conversion logic will already exist, so will not have to be written again or maintained separately. The advantage of doing this is that it makes the MAPI controller trivial, even if it doesn't go away completely.

Reply:
Having different methods across the platform may be more difficult to support, but, I don't think you suffer the same pain we do having to rebase/update many different platforms each release, so, perhaps this is ok to have the one off, this is different in OpenStack approach.

I assume the CPMS, MachineSet etc stuff will still apply, just not the Machine logic? 
Does this mean the authoritative API stuff you aren't expecting to participate in?

I'm not sure what that will mean for some of the goals here, but, for example, our intention is to eventually stop running Machine API operator, which currently deploys the OpenStack MAPI controller, so what would your plan be there?

Reply:
This is likely not worth going into in more detail yet until we know how the rest of the sync process will work.